pkg/cmd/pulumi/operations/preview.go

```diff
@@ -313,14 +316,28 @@ func NewPreviewCmd() *cobra.Command {
 		Args: cmdArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			ctx := cmd.Context()
+			ws := pkgWorkspace.Instance
 
-			err := validateAttachDebuggerFlag(attachDebugger)
+			proj, root, err := readProjectForUpdate(ws, client)
 			if err != nil {
 				return err
 			}
 
+			// Retrieve language runtime metadata async so we can continue with the other work in the meantime.
+			waitForMetadata := make(chan map[string]string, 1)
+			go func() {
+				env, err := metadata.GetLanguageRuntimeMetadata(root, proj)
+				if err != nil {
+					logging.V(1).Infof("Could not retrieve language runtime metadata: %s", err)
+				}
+				waitForMetadata <- env
+			}()
```

Comment:
If you use the promise library this will 
A) Hide the channel details
B) Let you correctly handle cancellation when waiting for the metadata